isis/src/base/objs/XmlToJson/XmlToJson.cpp

```diff
@@ -0,0 +1,212 @@
+  }
+  else {
+    // Just add element and its value
+    // <tag>value</tag>
+    if (!element.text().isEmpty()) {
+      newJson[element.tagName().toStdString()] = element.text().toStdString();
+    }
+    else {
+      // <tag /> no value case
+      newJson[element.tagName().toStdString()];
+    }
+    return newJson;
+  }
+  return newJson;
+}
+
+
+/**
+ * Not intended to be used directly. Intended to be used by xmlToJson to convert 
+ * an input XML document to JSON. 
+ *  
+ * This function does the following conversions: 
+ *  
+ * XML: <a><b>val1</b><c>val2</c></a> 
+ * JSON: a : {b: val1, c: val2}
+ *  
+ * XML: <a> <first>value1</first> </a> <a> <second>value2</second></a>
+ * JSON: a: [ {first:value1, second:value2} ]
+ *  
+ * XML: <a>val1</a><a>val2</a> 
+ * JSON: a:[val1, val2]
+ *  
+ * @param element A QDomElement representing the whole or some subset of a QDomDocument
+ * @param output A JSON object constructed from XML input.
+ * 
+ * @return ordered_json 
+ */
+ordered_json convertXmlToJson(QDomElement& element, ordered_json& output) {
+  int i = 0;
+  while (!element.isNull()) {
+    QDomElement next = element.firstChildElement();
+    if (next.isNull()){
+      ordered_json converted = convertLastChildNodeToJson(element);
+      // Simple case with no repeated tags at the same level
+      if (!output.contains(element.tagName().toStdString())){
+        output.update(converted);
+      }
+      else {
+        // There is a repeated tag at the same level in the XML, i.e: <a>val1</a><a>val2</a>. 
+        // Translated json goal: a:[val1, val2]
+        // If the converted json has an array already, append, else make it an array  
+        if (output[element.tagName().toStdString()].is_array()) {
+          output[element.tagName().toStdString()].push_back(converted[element.tagName().toStdString()]);
+        }
+        else {
+          output[element.tagName().toStdString()] = {output[element.tagName().toStdString()], converted[element.tagName().toStdString()]};
+        }
+      }
+    }
+    else {
+      // If there is already an element with this tag name at any level besides the same one,
+      // add it to a list rather than 
+      // overwriting. This is the following situation:
+      // XML: <a> <first>value1</first> </a> <a> <second>value2</second></a>
+      // JSON: a: [ {first:value1, second:value2} ]
+      if (output.contains(element.tagName().toStdString())) {
+        
+        // If it's an array already, append, else make it an array
+        if (output[element.tagName().toStdString()].is_array()) {
+          ordered_json temporaryJson;
+          convertXmlToJson(next, temporaryJson);
+          output[element.tagName().toStdString()].push_back(temporaryJson);
+        }
+        else {
+          ordered_json temporaryJson;
+          convertXmlToJson(next, temporaryJson);
+          output[element.tagName().toStdString()] = {output[element.tagName().toStdString()], temporaryJson};
+        }
```

Comment:
Same as above, this is much cleaner if you only have the conversion and append code once.